File: ex10_dev_kit/examples/get_version_info.py
#############################################################################
#                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      #
#                                                                           #
# This source code is the property of Impinj, Inc. Your use of this source  #
# code in whole or in part is subject to your applicable license terms      #
# from Impinj.                                                              #
# Contact support@impinj.com for a copy of the applicable Impinj license    #
# terms.                                                                    #
#                                                                           #
# (c) Copyright 2019 - 2021 Impinj, Inc. All rights reserved.               #
#                                                                           #
#############################################################################
""" Test the VersionInfo class """

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)

import argparse
import sys

from py2c_interface.py2c_python_wrapper import *


def _print_bootloader_version(py2c):
    """
    Print the bootloader version.
    Note that this will reset the Ex10 device to enter the bootloader.
    It will reset the Ex10 device a second time and attempt to start the
    Application.
    """
    ver_info = (c_char * VERSION_STRING_SIZE)()
    ex10_version = py2c.get_ex10_version()
    ex10_version.get_bootloader_info(ver_info, VERSION_STRING_SIZE)
    print(ver_info.value.decode("utf-8"))


def _print_application_version(py2c, ex10_protocol):
    """ Print the application version """
    # Attempt to reset into the Application
    ex10_protocol.reset(Status.Application)

    ver_info = (c_char * VERSION_STRING_SIZE)()
    image_validity = ImageValidityFields(1, 1)
    remain_reason = RemainReasonFields()
    remain_reason.remain_reason = 0
    ex10_version = py2c.get_ex10_version()
    ex10_version.get_application_info(ver_info, VERSION_STRING_SIZE, pointer(image_validity), pointer(remain_reason))
    print(ver_info.value.decode("utf-8"))

    if (image_validity.image_valid_marker) and not (image_validity.image_non_valid_marker):
        print("Application image VALID\n")
    else:
        print("Application image INVALID\n")

    reason_c_string = py2c.get_ex10_helpers().get_remain_reason_string(remain_reason.remain_reason)
    reason_string = (c_char * REMAIN_REASON_STRING_MAX_SIZE)()
    ctypes.memmove(reason_string, reason_c_string, REMAIN_REASON_STRING_MAX_SIZE)
    print("Remain in bootloader reason: {}".format(reason_string.value.decode("utf-8")))


def _print_device_version(py2c):
    """ Print the device version """
    ex10_version = py2c.get_ex10_version()
    dev_info = ex10_version.get_device_info()
    print(dev_info.decode("utf-8"))


def _print_sku(py2c, ex10_protocol):
    """ Print the device SKU """
    # Attempt to reset into the Application
    ex10_protocol.reset(Status.Application)

    ex10_version = py2c.get_ex10_version()
    sku_val = ex10_version.get_sku()
    if sku_val == ProductSku.SkuUnknown:
        print("SKU: Unknown\n")
    elif sku_val == ProductSku.SkuE310:
        print("SKU: E310\n")
    elif sku_val == ProductSku.SkuE510:
        print("SKU: E510\n")
    elif sku_val == ProductSku.SkuE710:
        print("SKU: E710\n")
    else:
        print("SKU: Unknown\n")


def main():
    """ Run the get_version_info() with user supplied parameters """

    parser = argparse.ArgumentParser(description='Get Version example')

    parser.add_argument('-a', '--application', action='store_true',
                        help='Print the application version information')

    parser.add_argument('-b', '--bootloader', action='store_true',
                        help='Print the Bootloader version information '
                        '(this will reset the device into the bootloader to'
                        ' retrieve the bootloader version and reset back to'
                        ' the application)')

    parser.add_argument('-d', '--device', action='store_true',
                        help='Print the Ex10 device version info')

    parser.add_argument('-s', '--sku', action='store_true',
                        help='Print the Ex10 SKU info')

    args = parser.parse_args()

    # If no argument provided, set all to True and print all
    if len(sys.argv) == 1:
        for arg in vars(args):
            setattr(args, arg, True)

    py2c = Ex10Py2CWrapper()
    ex10_protocol = py2c.ex10_bootloader_board_setup(BOOTLOADER_SPI_CLOCK_HZ)

    try:
        if args.application:
            _print_application_version(py2c, ex10_protocol)

        if args.bootloader:
            _print_bootloader_version(py2c)

        if args.device:
            _print_device_version(py2c)

        if args.sku:
            _print_sku(py2c, ex10_protocol)

    finally:
        py2c.ex10_typical_board_teardown()


if __name__ == "__main__":
    main()

File: ex10_c_dev_kit/examples/get_version_info.c
/*****************************************************************************
 *                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      *
 *                                                                           *
 * This source code is the property of Impinj, Inc. Your use of this source  *
 * code in whole or in part is subject to your applicable license terms      *
 * from Impinj.                                                              *
 * Contact support@impinj.com for a copy of the applicable Impinj license    *
 * terms.                                                                    *
 *                                                                           *
 * (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               *
 *                                                                           *
 *****************************************************************************/

#include <assert.h>
#include <stdio.h>

#include "calibration.h"
#include "ex10_api/application_registers.h"
#include "ex10_api/board_init.h"
#include "ex10_api/bootloader_registers.h"
#include "ex10_api/ex10_helpers.h"
#include "ex10_api/ex10_ops.h"
#include "ex10_api/trace.h"
#include "ex10_api/version_info.h"


#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))

static void calibration_version_info(struct Ex10Protocol const* ex10_protocol)
{
    // Attempt to reset into the Application
    ex10_protocol->reset(Application);

    printf("Calibration version: ");

    if (ex10_protocol->get_running_location() == Application)
    {
        struct Ex10Calibration const* calibration =
            get_ex10_calibration(ex10_protocol);

        printf("%u\n", calibration->get_cal_version());
    }
    else
    {
        // In the bootloader,
        // the calibration cannot be read via the Read command.
        printf("Unknown\n");
    }
}

static void sku_info(struct Ex10Protocol const* ex10_protocol)
{
    // Attempt to reset into the Application
    ex10_protocol->reset(Application);

    // Read the sku back from the device
    uint16_t sku_val = get_ex10_version()->get_sku();

    printf("SKU: ");
    switch (sku_val)
    {
        case SkuE310:
            printf("E310\n");
            break;
        case SkuE510:
            printf("E510\n");
            break;
        case SkuE710:
            printf("E710\n");
            break;
        case SkuE910:
            printf("E910\n");
            break;
        case SkuUnknown:
        default:
            printf("Unknown\n");
            break;
    }
}

static void device_info(void)
{
    const char* dev_info = get_ex10_version()->get_device_info();
    printf("%s\n", dev_info);
}

static void app_version_info(struct Ex10Protocol const* ex10_protocol)
{
    char                       ver_info[VERSION_STRING_SIZE];
    struct ImageValidityFields image_validity;
    struct RemainReasonFields  remain_reason;

    // Attempt to reset into the Application
    ex10_protocol->reset(Application);

    get_ex10_version()->get_application_info(
        ver_info, sizeof(ver_info), &image_validity, &remain_reason);
    printf("%s\n", ver_info);

    if ((image_validity.image_valid_marker) &&
        !(image_validity.image_non_valid_marker))
    {
        printf("Application image VALID\n");
    }
    else
    {
        printf("Application image INVALID\n");
    }

    printf("Remain in bootloader reason: %s\n",
           get_ex10_helpers()->get_remain_reason_string(
               remain_reason.remain_reason));
}

static void bootloader_version_info(void)
{
    char ver_info[VERSION_STRING_SIZE];

    get_ex10_version()->get_bootloader_info(ver_info, sizeof(ver_info));
    printf("%s\n", ver_info);
}

static void print_help(void)
{
    printf("No args given. Accepted arguments are...\n");
    printf("a: returns application info,\n");
    printf("b: returns bootloader info,\n");
    printf("c: returns calibration info\n");
    printf("d: returns device info,\n");
    printf("s: returns sku\n");
}

int main(int argc, char* argv[])
{
    tracepoint(pi_ex10sdk, EXEC_start, __FILE__);
    int result = 0;

    // Initialize version_list to print all.
    // The default set of versions to print when no arguments are specified.
    char         version_list[]    = {'a', 'b', 'c', 'd', 's'};
    size_t const version_list_size = ARRAY_SIZE(version_list);
    size_t       param_length      = version_list_size;

    if (argc > 1)
    {
        // Command arguments specified.
        // argc is all params and 1 count for the script itself
        param_length = argc - 1;
        if (param_length > version_list_size)
        {
            fprintf(stderr,
                    "warning: command line argument count %zu "
                    "exceeds allowed count %zu, extra arguments dropped\n",
                    param_length,
                    version_list_size);
            param_length = version_list_size;
        }

        for (size_t i = 0; i < param_length; i++)
        {
            // skip the script name in argv
            version_list[i] = *(argv[i + 1]);
        }
    }

    struct Ex10Protocol const* ex10_protocol =
        ex10_bootloader_board_setup(BOOTLOADER_SPI_CLOCK_HZ);

    for (size_t i = 0; i < param_length; i++)
    {
        // The argument input is the image file
        char version_arg = version_list[i];

        switch (version_arg)
        {
            case 'a':
                app_version_info(ex10_protocol);
                break;
            case 'b':
                bootloader_version_info();
                break;
            case 'c':
                calibration_version_info(ex10_protocol);
                break;
            case 'd':
                device_info();
                break;
            case 's':
                sku_info(ex10_protocol);
                break;
            default:
                print_help();
                return 1;
                break;
        }
    }

    ex10_typical_board_teardown();

    tracepoint(pi_ex10sdk, EXEC_end, __FILE__);
    return result;
}
